module/state_synchronization/serializer.go

```diff
@@ -0,0 +1,97 @@
+package state_synchronization
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/ipfs/go-cid"
+
+	"github.com/onflow/flow-go/model/encoding"
+	"github.com/onflow/flow-go/network"
+)
+
+// header codes to distinguish between different types of data
+const (
+	CodeRecursiveCIDs = iota
+	CodeExecutionData
+)
+
+func getCode(v interface{}) byte {
+	switch v.(type) {
+	case *ExecutionData:
+		return CodeExecutionData
+	case []cid.Cid:
+		return CodeRecursiveCIDs
+	default:
+		panic(fmt.Sprintf("invalid type for interface: %T", v))
+	}
+}
+
+func getPrototype(code byte) interface{} {
+	switch code {
+	case CodeExecutionData:
+		return &ExecutionData{}
+	case CodeRecursiveCIDs:
+		return &[]cid.Cid{}
+	default:
+		panic(fmt.Sprintf("invalid code: %v", code))
+	}
+}
+
+type reader interface {
+	io.ByteReader
+	io.Reader
+}
+
+type writer interface {
+	io.ByteWriter
+	io.Writer
+}
+
+// serializer encapulates the serialization and deserialization of data
+type serializer struct {
+	codec      encoding.Codec
+	compressor network.Compressor
+}
+
+func (s *serializer) serialize(w writer, v interface{}) error {
+	if err := w.WriteByte(getCode(v)); err != nil {
```

Comment:
So, we don't do the splitting of data into blocks until after it has been serialized. As you see, the `serializer` takes in the `BlockWriter` as input and writes the serialized data to the `BlockWriter`. Therefore, it doesn't really make sense to talk about blocks being shifted by one byte because from the perspective of the `BlockWriter`, it is just chunking an arbitrary sequence of bytes into 1mb blocks.

One way to think of it is to just view the serializer as implementing some custom serialization algorithm. Then, all we are doing is we are taking the serialized representation of an object, and then chunking it into blocks, which is precisely what we are supposed to.

Under the hood, this "custom serialization algorithm" is just adding an extra byte to the beginning of the CBOR serialization of the object, but this is just an implementation detail really.

Reply:
Furthermore, there's unfortunately no way to include "metadata" in a block (at least if we're using the IPFS implementation). The only information associated with a block is the raw bytes themselves (from which a CID may be derived), nothing more and nothing less.